posthog/temporal/batch_exports/s3_batch_export.py

```diff
@@ -416,42 +419,85 @@ async def __aexit__(self, exc_type, exc_value, traceback) -> bool:
         return False
 
 
+@dataclasses.dataclass
+class S3InsertInputs:
+    """Inputs for S3 exports."""
+
+    # TODO: do _not_ store credentials in temporal inputs. It makes it very hard
+    # to keep track of where credentials are being stored and increases the
+    # attach surface for credential leaks.
+
+    bucket_name: str
+    region: str
+    prefix: str
+    team_id: int
+    data_interval_start: str | None
+    data_interval_end: str
+    aws_access_key_id: str | None = None
+    aws_secret_access_key: str | None = None
+    compression: str | None = None
+    exclude_events: list[str] | None = None
+    include_events: list[str] | None = None
+    encryption: str | None = None
+    kms_key_id: str | None = None
+    endpoint_url: str | None = None
+    # TODO: In Python 3.11, this could be a enum.StrEnum.
+    file_format: str = "JSONLines"
+    run_id: str | None = None
+    is_backfill: bool = False
+    batch_export_model: BatchExportModel | None = None
+    # TODO: Remove after updating existing batch exports
+    batch_export_schema: BatchExportSchema | None = None
+
+
 @dataclasses.dataclass
 class S3HeartbeatDetails(BatchExportRangeHeartbeatDetails):
     """This tuple allows us to enforce a schema on the Heartbeat details.
 
     Attributes:
         upload_state: State to continue a S3MultiPartUpload when activity execution resumes.
+        files_uploaded: The number of files we have uploaded so far
```

Comment:
probably worth checking I've done this right. Not sure if there's an easy way to test out resuming an activity from a previous heartbeat?

Reply:
There is, several tests cover this. Important to check what happens with Parquet which could not be valid if we do not clean-up properly when retrying.

Reply:
For example, this test feeds a hand-crafted heartbeat to assert how the batch export completes a run: https://github.com/PostHog/posthog/blob/master/posthog/temporal/tests/batch_exports/test_bigquery_batch_export_workflow.py#L684

Reply:
Another semi-related example: This test captures the contents of heartbeats and asserts what's in them: https://github.com/PostHog/posthog/blob/master/posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py#L1458